File: typescript/test/encoder.test.ts
import { describe, it, expect } from 'vitest';
import { EncoderBase32Crockford } from '../src/encoder';

describe('EncoderBase32Crockford', () => {
  describe('encode', () => {
    it('should encode zero', () => {
      expect(EncoderBase32Crockford.encode(0)).toBe('0');
    });

    it('should encode small numbers', () => {
      expect(EncoderBase32Crockford.encode(1)).toBe('1');
      expect(EncoderBase32Crockford.encode(31)).toBe('z');
      expect(EncoderBase32Crockford.encode(32)).toBe('10');
    });

    it('should encode large numbers', () => {
      expect(EncoderBase32Crockford.encode(1000)).toBe('z8');
      expect(EncoderBase32Crockford.encode(123456)).toBe('3rj0');
    });

    it('should pad when requested', () => {
      const encoded = EncoderBase32Crockford.encode(123, true);
      expect(encoded).toHaveLength(13);
      expect(encoded).toBe('000000000003v');
    });

    it('should not pad by default', () => {
      const encoded = EncoderBase32Crockford.encode(123);
      expect(encoded.length).toBeLessThan(13);
      expect(encoded).toBe('3v');
    });

    it('should handle timestamp-like numbers', () => {
      const timestamp = Date.now();
      const encoded = EncoderBase32Crockford.encode(timestamp);
      expect(encoded).toBeTruthy();
      expect(encoded.length).toBeGreaterThan(0);
    });

    it('should throw on negative numbers', () => {
      expect(() => EncoderBase32Crockford.encode(-1)).toThrow();
    });

    it('should throw on non-finite numbers', () => {
      expect(() => EncoderBase32Crockford.encode(NaN)).toThrow();
      expect(() => EncoderBase32Crockford.encode(Infinity)).toThrow();
    });
  });

  describe('decode', () => {
    it('should decode zero', () => {
      expect(EncoderBase32Crockford.decode('0')).toBe(0);
    });

    it('should decode small numbers', () => {
      expect(EncoderBase32Crockford.decode('1')).toBe(1);
      expect(EncoderBase32Crockford.decode('z')).toBe(31);
      expect(EncoderBase32Crockford.decode('10')).toBe(32);
    });

    it('should decode large numbers', () => {
      expect(EncoderBase32Crockford.decode('z8')).toBe(1000);
      expect(EncoderBase32Crockford.decode('3rj0')).toBe(123456);
    });

    it('should be case-insensitive', () => {
      expect(EncoderBase32Crockford.decode('ABC')).toBe(
        EncoderBase32Crockford.decode('abc')
      );
      expect(EncoderBase32Crockford.decode('XyZ')).toBe(
        EncoderBase32Crockford.decode('xyz')
      );
    });

    it('should map lookalike characters (O → 0)', () => {
      expect(EncoderBase32Crockford.decode('O')).toBe(0);
      expect(EncoderBase32Crockford.decode('o')).toBe(0);
      expect(EncoderBase32Crockford.decode('0')).toBe(0);
    });

    it('should map lookalike characters (I/L → 1)', () => {
      expect(EncoderBase32Crockford.decode('I')).toBe(1);
      expect(EncoderBase32Crockford.decode('i')).toBe(1);
      expect(EncoderBase32Crockford.decode('L')).toBe(1);
      expect(EncoderBase32Crockford.decode('l')).toBe(1);
      expect(EncoderBase32Crockford.decode('1')).toBe(1);
    });

    it('should throw on empty string', () => {
      expect(() => EncoderBase32Crockford.decode('')).toThrow();
    });

    it('should throw on invalid characters', () => {
      expect(() => EncoderBase32Crockford.decode('abc-def')).toThrow();
      expect(() => EncoderBase32Crockford.decode('abc def')).toThrow();
      expect(() => EncoderBase32Crockford.decode('abc_def')).toThrow();
    });
  });

  describe('roundtrip encoding/decoding', () => {
    it('should encode and decode to same value', () => {
      const testCases = [0, 1, 31, 32, 100, 1000, 123456, Date.now()];

      testCases.forEach((num) => {
        const encoded = EncoderBase32Crockford.encode(num);
        const decoded = EncoderBase32Crockford.decode(encoded);
        expect(decoded).toBe(num);
      });
    });

    it('should handle padded encoding roundtrip', () => {
      const num = 12345;
      const encoded = EncoderBase32Crockford.encode(num, true);
      const decoded = EncoderBase32Crockford.decode(encoded);
      expect(decoded).toBe(num);
    });
  });

  describe('encodeBigInt', () => {
    it('should encode zero', () => {
      expect(EncoderBase32Crockford.encodeBigInt(0n)).toBe('0');
    });

    it('should encode small BigInt values', () => {
      expect(EncoderBase32Crockford.encodeBigInt(1n)).toBe('1');
      expect(EncoderBase32Crockford.encodeBigInt(31n)).toBe('z');
      expect(EncoderBase32Crockford.encodeBigInt(32n)).toBe('10');
    });

    it('should encode values beyond MAX_SAFE_INTEGER', () => {
      // 2^53 is beyond MAX_SAFE_INTEGER
      const beyondSafe = BigInt(Number.MAX_SAFE_INTEGER) + 1n;
      const encoded = EncoderBase32Crockford.encodeBigInt(beyondSafe);
      expect(encoded).toBeTruthy();
      expect(encoded.length).toBeGreaterThan(10);
    });

    it('should encode full 64-bit values', () => {
      // Max 64-bit unsigned value: 2^64 - 1 = 18446744073709551615
      const max64bit = 18446744073709551615n;
      const encoded = EncoderBase32Crockford.encodeBigInt(max64bit);
      // 64 bits needs about 13 base32 chars (65 bits / 5 bits per char)
      expect(encoded).toBe('fzzzzzzzzzzzz');
    });

    it('should pad when requested', () => {
      const encoded = EncoderBase32Crockford.encodeBigInt(123n, true);
      expect(encoded).toHaveLength(13);
      expect(encoded).toBe('000000000003v');
    });

    it('should throw on negative BigInt', () => {
      expect(() => EncoderBase32Crockford.encodeBigInt(-1n)).toThrow();
    });
  });

  describe('decodeBigInt', () => {
    it('should decode zero', () => {
      expect(EncoderBase32Crockford.decodeBigInt('0')).toBe(0n);
    });

    it('should decode small values', () => {
      expect(EncoderBase32Crockford.decodeBigInt('1')).toBe(1n);
      expect(EncoderBase32Crockford.decodeBigInt('z')).toBe(31n);
      expect(EncoderBase32Crockford.decodeBigInt('10')).toBe(32n);
    });

    it('should decode values beyond MAX_SAFE_INTEGER', () => {
      const beyondSafe = BigInt(Number.MAX_SAFE_INTEGER) + 1n;
      const encoded = EncoderBase32Crockford.encodeBigInt(beyondSafe);
      const decoded = EncoderBase32Crockford.decodeBigInt(encoded);
      expect(decoded).toBe(beyondSafe);
    });

    it('should decode full 64-bit values', () => {
      const max64bit = 18446744073709551615n;
      const encoded = EncoderBase32Crockford.encodeBigInt(max64bit);
      const decoded = EncoderBase32Crockford.decodeBigInt(encoded);
      expect(decoded).toBe(max64bit);
    });

    it('should be case-insensitive', () => {
      expect(EncoderBase32Crockford.decodeBigInt('ABC')).toBe(
        EncoderBase32Crockford.decodeBigInt('abc')
      );
    });
  });

  describe('BigInt roundtrip', () => {
    it('should encode and decode to same BigInt value', () => {
      const testCases = [0n, 1n, 31n, 32n, 100n, 1000n, 123456n, BigInt(Date.now())];

      testCases.forEach((num) => {
        const encoded = EncoderBase32Crockford.encodeBigInt(num);
        const decoded = EncoderBase32Crockford.decodeBigInt(encoded);
        expect(decoded).toBe(num);
      });
    });

    it('should handle values beyond safe integer range', () => {
      const testCases = [
        BigInt(Number.MAX_SAFE_INTEGER) + 1n,
        BigInt(Number.MAX_SAFE_INTEGER) * 2n,
        18446744073709551615n, // max 64-bit
      ];

      testCases.forEach((num) => {
        const encoded = EncoderBase32Crockford.encodeBigInt(num);
        const decoded = EncoderBase32Crockford.decodeBigInt(encoded);
        expect(decoded).toBe(num);
      });
    });
  });

  describe('secureRandomLong', () => {
    it('should generate positive BigInt values', () => {
      for (let i = 0; i < 100; i++) {
        const random = EncoderBase32Crockford.secureRandomLong();
        expect(typeof random).toBe('bigint');
        expect(random).toBeGreaterThanOrEqual(0n);
      }
    });

    it('should generate different values', () => {
      const values = new Set<bigint>();
      for (let i = 0; i < 100; i++) {
        values.add(EncoderBase32Crockford.secureRandomLong());
      }
      // Should have generated many unique values (allow some collisions)
      expect(values.size).toBeGreaterThan(90);
    });

    it('should generate full 64-bit entropy', () => {
      // With 64-bit entropy, we should see values that exceed 53-bit MAX_SAFE_INTEGER
      const samples = 100;
      let sawValueBeyondSafe = false;
      const threshold = BigInt(Number.MAX_SAFE_INTEGER);

      for (let i = 0; i < samples; i++) {
        const random = EncoderBase32Crockford.secureRandomLong();
        if (random > threshold) {
          sawValueBeyondSafe = true;
          break;
        }
      }

      // With uniform 64-bit distribution, ~50% of values should exceed 2^53
      // So we should definitely see at least one in 100 samples
      expect(sawValueBeyondSafe).toBe(true);
    });

    it('should produce values that encode to full 13-char width', () => {
      // Values using full 64 bits should encode to 13 chars (not just 11 for 53-bit)
      const samples = 100;
      let saw13CharEncoding = false;

      for (let i = 0; i < samples; i++) {
        const random = EncoderBase32Crockford.secureRandomLong();
        const encoded = EncoderBase32Crockford.encodeBigInt(random);
        if (encoded.length === 13) {
          saw13CharEncoding = true;
          break;
        }
      }

      expect(saw13CharEncoding).toBe(true);
    });
  });

  describe('isValid', () => {
    it('should validate correct Base32 strings', () => {
      expect(EncoderBase32Crockford.isValid('0')).toBe(true);
      expect(EncoderBase32Crockford.isValid('1')).toBe(true);
      expect(EncoderBase32Crockford.isValid('abc123')).toBe(true);
      expect(EncoderBase32Crockford.isValid('1fvszawr42tve3gxvx9900')).toBe(true);
    });

    it('should reject invalid strings', () => {
      expect(EncoderBase32Crockford.isValid('')).toBe(false);
      expect(EncoderBase32Crockford.isValid('abc-def')).toBe(false);
      expect(EncoderBase32Crockford.isValid('abc def')).toBe(false);
      expect(EncoderBase32Crockford.isValid('abc_def')).toBe(false);
    });

    it('should accept lookalike characters', () => {
      expect(EncoderBase32Crockford.isValid('OIL')).toBe(true); // O→0, I→1, L→1
      expect(EncoderBase32Crockford.isValid('oil')).toBe(true);
    });
  });
});

File: rust/tests/encoder_test.rs
use ashid::EncoderBase32Crockford;

#[test]
fn encode_zero() {
    assert_eq!(EncoderBase32Crockford::encode(0, false), "0");
}

#[test]
fn encode_small_numbers() {
    assert_eq!(EncoderBase32Crockford::encode(1, false), "1");
    assert_eq!(EncoderBase32Crockford::encode(31, false), "z");
    assert_eq!(EncoderBase32Crockford::encode(32, false), "10");
}

#[test]
fn encode_large_numbers() {
    assert_eq!(EncoderBase32Crockford::encode(1000, false), "z8");
    assert_eq!(EncoderBase32Crockford::encode(123456, false), "3rj0");
}

#[test]
fn encode_padded() {
    let encoded = EncoderBase32Crockford::encode(123, true);
    assert_eq!(encoded.len(), 13);
    assert_eq!(encoded, "000000000003v");
}

#[test]
fn encode_unpadded_default() {
    let encoded = EncoderBase32Crockford::encode(123, false);
    assert!(encoded.len() < 13);
    assert_eq!(encoded, "3v");
}

#[test]
fn encode_zero_padded() {
    assert_eq!(EncoderBase32Crockford::encode(0, true), "0000000000000");
}

#[test]
fn encode_full_64bit_max() {
    // 2^64 - 1 = 18446744073709551615
    let encoded = EncoderBase32Crockford::encode(u64::MAX, false);
    assert_eq!(encoded, "fzzzzzzzzzzzz");
}

#[test]
fn encode_full_64bit_padded() {
    let encoded = EncoderBase32Crockford::encode(u64::MAX, true);
    assert_eq!(encoded.len(), 13);
    assert_eq!(encoded, "fzzzzzzzzzzzz");
}

#[test]
fn decode_zero() {
    assert_eq!(EncoderBase32Crockford::decode("0").unwrap(), 0);
}

#[test]
fn decode_small_numbers() {
    assert_eq!(EncoderBase32Crockford::decode("1").unwrap(), 1);
    assert_eq!(EncoderBase32Crockford::decode("z").unwrap(), 31);
    assert_eq!(EncoderBase32Crockford::decode("10").unwrap(), 32);
}

#[test]
fn decode_large_numbers() {
    assert_eq!(EncoderBase32Crockford::decode("z8").unwrap(), 1000);
    assert_eq!(EncoderBase32Crockford::decode("3rj0").unwrap(), 123456);
}

#[test]
fn decode_case_insensitive() {
    assert_eq!(
        EncoderBase32Crockford::decode("ABC").unwrap(),
        EncoderBase32Crockford::decode("abc").unwrap()
    );
    assert_eq!(
        EncoderBase32Crockford::decode("XyZ").unwrap(),
        EncoderBase32Crockford::decode("xyz").unwrap()
    );
}

#[test]
fn decode_lookalike_o_to_zero() {
    assert_eq!(EncoderBase32Crockford::decode("O").unwrap(), 0);
    assert_eq!(EncoderBase32Crockford::decode("o").unwrap(), 0);
    assert_eq!(EncoderBase32Crockford::decode("0").unwrap(), 0);
}

#[test]
fn decode_lookalike_i_l_to_one() {
    assert_eq!(EncoderBase32Crockford::decode("I").unwrap(), 1);
    assert_eq!(EncoderBase32Crockford::decode("i").unwrap(), 1);
    assert_eq!(EncoderBase32Crockford::decode("L").unwrap(), 1);
    assert_eq!(EncoderBase32Crockford::decode("l").unwrap(), 1);
    assert_eq!(EncoderBase32Crockford::decode("1").unwrap(), 1);
}

#[test]
fn decode_lookalike_u_to_v() {
    assert_eq!(EncoderBase32Crockford::decode("U").unwrap(), 27);
    assert_eq!(EncoderBase32Crockford::decode("u").unwrap(), 27);
    assert_eq!(EncoderBase32Crockford::decode("V").unwrap(), 27);
    assert_eq!(EncoderBase32Crockford::decode("v").unwrap(), 27);
}

#[test]
fn decode_empty_string_errors() {
    assert!(EncoderBase32Crockford::decode("").is_err());
}

#[test]
fn decode_invalid_chars_error() {
    assert!(EncoderBase32Crockford::decode("abc-def").is_err());
    assert!(EncoderBase32Crockford::decode("abc def").is_err());
    assert!(EncoderBase32Crockford::decode("abc_def").is_err());
}

#[test]
fn roundtrip_basic() {
    for &n in &[0u64, 1, 31, 32, 100, 1000, 123_456, 1_700_000_000_000] {
        let encoded = EncoderBase32Crockford::encode(n, false);
        let decoded = EncoderBase32Crockford::decode(&encoded).unwrap();
        assert_eq!(decoded, n);
    }
}

#[test]
fn roundtrip_padded() {
    let n = 12345u64;
    let encoded = EncoderBase32Crockford::encode(n, true);
    let decoded = EncoderBase32Crockford::decode(&encoded).unwrap();
    assert_eq!(decoded, n);
}

#[test]
fn roundtrip_full_64bit() {
    for &n in &[u64::MAX, u64::MAX - 1, 1u64 << 63, (1u64 << 53) + 1] {
        let encoded = EncoderBase32Crockford::encode(n, false);
        let decoded = EncoderBase32Crockford::decode(&encoded).unwrap();
        assert_eq!(decoded, n);
    }
}

#[test]
fn secure_random_long_generates_different_values() {
    let mut values = std::collections::HashSet::new();
    for _ in 0..100 {
        values.insert(EncoderBase32Crockford::secure_random_long());
    }
    assert!(
        values.len() > 90,
        "expected mostly unique values, got {}",
        values.len()
    );
}

#[test]
fn secure_random_long_full_64bit_entropy() {
    // With 64-bit entropy, ~50% of values should exceed 2^53
    let threshold = (1u64 << 53) - 1;
    let saw_beyond = (0..100).any(|_| EncoderBase32Crockford::secure_random_long() > threshold);
    assert!(
        saw_beyond,
        "expected to see at least one value beyond 53-bit safe range in 100 samples"
    );
}

#[test]
fn secure_random_long_produces_13_char_encoding() {
    let saw_13 = (0..100).any(|_| {
        let r = EncoderBase32Crockford::secure_random_long();
        EncoderBase32Crockford::encode(r, false).len() == 13
    });
    assert!(
        saw_13,
        "expected at least one value encoding to full 13 chars"
    );
}

#[test]
fn is_valid_accepts_correct_strings() {
    assert!(EncoderBase32Crockford::is_valid("0"));
    assert!(EncoderBase32Crockford::is_valid("1"));
    assert!(EncoderBase32Crockford::is_valid("abc123"));
    assert!(EncoderBase32Crockford::is_valid("1fvszawr42tve3gxvx9900"));
}

#[test]
fn is_valid_rejects_invalid_strings() {
    assert!(!EncoderBase32Crockford::is_valid(""));
    assert!(!EncoderBase32Crockford::is_valid("abc-def"));
    assert!(!EncoderBase32Crockford::is_valid("abc def"));
    assert!(!EncoderBase32Crockford::is_valid("abc_def"));
}

#[test]
fn is_valid_accepts_lookalikes() {
    assert!(EncoderBase32Crockford::is_valid("OIL")); // O→0, I→1, L→1
    assert!(EncoderBase32Crockford::is_valid("oil"));
}
